### ISSUE TITLE
Array destructuring transpiles into code that is incompatible with IE11

### ISSUE THREAD

**Issue by rhpijnacker** [babel/babel#11195]
## Bug Report

- [x] I would like to work on a fix!

**Current Behavior**
Transpiling the code below generates code that uses _Symbol_, which does not run on IE11, even though I have `@babel/env` set to `ie: 11`. 

**Input Code**
```js
const oneTwoTree = () => [1, 2, 3];
const [a, b, c] = oneTwoTree();
```

**Output Code**
```js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var oneTwoTree = function oneTwoTree() {
  return [1, 2, 3];
};

var _oneTwoTree = oneTwoTree(),
    _oneTwoTree2 = _slicedToArray(_oneTwoTree, 3),
    a = _oneTwoTree2[0],
    b = _oneTwoTree2[1],
    c = _oneTwoTree2[2];
```

**Expected behavior/code**
Transpiled code that runs on IE11.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
module.exports = function(api) {                                                                                       
  return {                                                                                                             
    presets: [                                                                                                         
      [                                                                                                                
        '@babel/env',                                                                                                  
        {                                                                                                              
          modules: 'amd',                                                                                              
          targets: {                                                                                                   
            ie: '11',                                                                                                  
            safari: '11'                                                                                               
          }                                                                                                            
        }                                                                                                              
      ],                                                                                                               
      ['@babel/react']                                                                                                 
    ],                                                                                                                 
    sourceMaps: 'both',                                                                                                
    sourceType: 'unambiguous'                                                                                          
  };                                                                                                                   
};                                                                                                                     
```

**Environment**

- Babel version(s): v7.8.4
- How you are using Babel: `cli`

**Comment by nicolo-ribaudo** [comment]
What we usually call "array destructuring" isn't really for arrays. It's for "iterables", which are "objects with a `Symbol.iterator` property". Arrays fall into this category.

This means that the transform should directly use `Symbol.iterator`, because `[a, b, c] = ...` is just syntax sugar for calling `Symbol.iterator`.

However, we have special-cased arrays for two reasons:
1. Most of the "array destructuring" is used with arrays
2. Old browsers like IE11 support arrays but don't have `Symbol.iterator`

Your example falls in the second category: if you run that code using IE11, _it works_.
You can simulate it in a modern browser, by replacing the body of `_iterableToArrayLimit` (which uses `Symbol`) with a `throw` statement:

```js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { throw new Error("Symbols are not supported!"); }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var oneTwoTree = function oneTwoTree() {
  return [1, 2, 3];
};

var _oneTwoTree = oneTwoTree(),
    _oneTwoTree2 = _slicedToArray(_oneTwoTree, 3),
    a = _oneTwoTree2[0],
    b = _oneTwoTree2[1],
    c = _oneTwoTree2[2];

console.log(a, b, c);
```

You can run it in any browser, and it will log `1, 2, 3`.

It works because `_slicedToArray(_oneTwoThree, 3)` first calls `_arrayWithHoles(_oneTwoThree)`. Since `_oneTwoThree` is an array, `Array.isArray(_oneTwoThree)` returns `true` and thus `_arrayWithHoles` returns `_oneTwoThree`, which is returned by `_slicedToArray` **without calling `_iterableToArrayLimit`**.

That said, Babel transforms are about syntax: different plugins use native functions not supported by some browsers, and you have to provide a polyfill (either manually, or with the `useBuiltIns` option of `@babel/preset-env`).

**Comment by rhpijnacker** [comment]
Ok, I get what you said above. I just now realize that I simplified my example to the point where it does not actually fail on IE11 anymore.

In the original code, it was not actually an array containing 1,2,3 , but rather a NodeList instance.
If you change the above example as follows:
```js
const oneTwoTree = () => document.body.childNodes;
const [a, b, c] = oneTwoTree();
```
it will actually fail with the message `'Symbol' is undefined`.

That leaves the question whether this is proper use of the destructuring assingment, since it is not a proper array... what do you think?

**Comment by nicolo-ribaudo** [comment]
You can now enable the `allowArrayLike` option of the destructuring transform, or the new `arrayLikeIsIterable` assumption that will be released in Babel 7.13.0 (https://github.com/babel/rfcs/pull/5).

**Comment by luwes** [comment]
Leaving this snippet here for anyone getting an error in IE11, hope it can help anyone.
It took pretty long for me to debug...

```
Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.
```

```js
function iterator() {
    var selfThis = this;
    var index = 0;
    var done;

    function Iterator() {}

    Iterator.prototype.next = function() {
        if (index > selfThis.length - 1) {
            done = true;
        }
        if (done) {
            return { value: undefined, done: true };
        }
        return { value: [index, selfThis[index++]], done: false };
    };
	
	// Many small polyfills leave out this method which Babel requires and otherwise throws the above error
    Iterator.prototype[Symbol.iterator] = function() {
        return this;
    };

    return new Iterator();
}

// Needed for: IE 11
if (!Array.prototype.entries) {
    Array.prototype.entries = iterator;
}

// Needed for: IE 11
if (!Array.prototype[Symbol.iterator]) {
    Array.prototype[Symbol.iterator] = iterator;
}
```
